src/main/java/software/amazon/cloudformation/metrics/MetricsPublisherImpl.java

```diff
@@ -57,86 +57,83 @@ public void publishExceptionMetric(final Instant timestamp,
+                    .build(),
+                Dimension.builder().name(Metric.DIMENSION_KEY_EXCEPTION_TYPE).value(e.getClass().toString()).build(),
+                Dimension.builder().name(Metric.DIMENSION_KEY_RESOURCE_TYPE).value(this.getResourceTypeName()).build(),
+                Dimension.builder().name(Metric.DIMENSION_KEY_HANDLER_ERROR_CODE).value(handlerErrorCode.name()).build()))
+            .build());
     }
 
     @Override
-    public void publishExceptionByErrorCodeMetric(final Instant timestamp,
-                                                  final Action action,
-                                                  final HandlerErrorCode handlerErrorCode,
-                                                  final boolean thrown) {
-        Map<String, String> dimensions = new HashMap<>();
-        dimensions.put(Metric.DIMENSION_KEY_ACTION_TYPE, action == null ? "NO_ACTION" : action.name());
-        dimensions.put(Metric.DIMENSION_KEY_HANDLER_ERROR_CODE, handlerErrorCode.name());
-
-        publishMetric(Metric.METRIC_NAME_HANDLER_EXCEPTION_BY_ERROR_CODE, dimensions, StandardUnit.COUNT, thrown ? 1.0 : 0.0,
-            timestamp);
-    }
-
-    public void publishExceptionCountMetric(final Instant timestamp, final Action action, final boolean thrown) {
-        Map<String, String> dimensions = new HashMap<>();
-        dimensions.put(Metric.DIMENSION_KEY_ACTION_TYPE, action == null ? "NO_ACTION" : action.name());
-
-        publishMetric(Metric.METRIC_NAME_HANDLER_EXCEPTION_BY_EXCEPTION_COUNT, dimensions, StandardUnit.COUNT, thrown ? 1.0 : 0.0,
-            timestamp);
+    public void publishExceptionByErrorCodeAndCountBulkMetrics(final Instant timestamp,
+                                                               final Action action,
+                                                               final HandlerErrorCode handlerErrorCode) {
+        Set<MetricDatum> bulkData = new HashSet<>();
+
+        // By Error Code dimensions
+
+        EnumSet.allOf(HandlerErrorCode.class).forEach(
+            errorCode -> bulkData.add(MetricDatum.builder().metricName(Metric.METRIC_NAME_HANDLER_EXCEPTION_BY_ERROR_CODE)
+                .unit(StandardUnit.COUNT).value(errorCode == handlerErrorCode ? 1.0 : 0.0)
+                .dimensions(Sets.newHashSet(
+                    Dimension.builder().name(Metric.DIMENSION_KEY_ACTION_TYPE).value(action == null ? "NO_ACTION" : action.name())
+                        .build(),
+                    Dimension.builder().name(Metric.DIMENSION_KEY_HANDLER_ERROR_CODE).value(errorCode.name()).build()))
+                .timestamp(timestamp).build()));
+
+        // By Count dimensions
+        bulkData.add(MetricDatum.builder().metricName(Metric.METRIC_NAME_HANDLER_EXCEPTION_BY_EXCEPTION_COUNT)
+            .unit(StandardUnit.COUNT).value(1.0).dimensions(Dimension.builder().name(Metric.DIMENSION_KEY_ACTION_TYPE)
```

Comment:
This should not be `1.0` all the time, otherwise, the metrics will be `1` for every data point. 

It should be `handlerErrorCode == null ? 0.0 : 1.0` , if there is error code, the it's exception.

Reply:
good catch!